core/src/main/java/org/apache/calcite/rel/rules/ProjectAggregateMergeRule.java

```diff
@@ -99,7 +100,8 @@ && kindCount(project.getProjects(), SqlKind.CASE) == 0) {
               && operands.get(1).getKind() == SqlKind.CAST
               && ((RexCall) operands.get(1)).operands.get(0).getKind()
               == SqlKind.INPUT_REF
-              && operands.get(2).getKind() == SqlKind.LITERAL) {
+              && operands.get(2).getKind() == SqlKind.LITERAL
+              && operands.get(2).getType().getFamily() == SqlTypeFamily.NUMERIC) {
```

Comment:
I wonder where the code `literal.getValueAs(BigDecimal.class)` (line 111) could work as [getValueAs method](https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/rex/RexLiteral.java#L1029)  has no case for "clazz == BigDecimal" ... May be not related to this specific case, but it seems there is also an issue on RexLiteral class.

Reply:
@clesaec 
I agree with you on this. I believe there could be a conversion for `BigDecimal` from `TIMESTAMP`.